java/ql/src/semmle/code/java/frameworks/apache/Lang.qll

```diff
@@ -25,3 +24,61 @@ class MethodApacheSerializationUtilsDeserialize extends Method {
+  result.(RefType).hasQualifiedName("java.util", "Locale")
+}
+
+private class SafeJoinArrayType extends PrimitiveType {
+  SafeJoinArrayType() { getName().regexpMatch("byte|double|float|int|long|short") }
+}
+
+private class ApacheStringUtilsTaintPreservingMethod extends TaintPreservingCallable {
+  ApacheStringUtilsTaintPreservingMethod() {
+    this.getDeclaringType().hasQualifiedName("org.apache.commons.lang3", "StringUtils") and
+    this.hasName([
+        "abbreviate", "abbreviateMiddle", "appendIfMissing", "appendIfMissingIgnoreCase",
+        "capitalize", "center", "chomp", "chop", "defaultIfBlank", "defaultIfEmpty",
+        "defaultString", "deleteWhitespace", "difference", "firstNonBlank", "firstNonEmpty",
+        "getBytes", "getCommonPrefix", "getDigits", "getIfBlank", "getIfEmpty", "join", "joinWith",
+        "left", "leftPad", "lowerCase", "mid", "normalizeSpace", "overlay", "prependIfMissing",
+        "prependIfMissingIgnoreCase", "remove", "removeAll", "removeEnd", "removeEndIgnoreCase",
+        "removeFirst", "removeIgnoreCase", "removePattern", "removeStart", "removeStartIgnoreCase",
+        "repeat", "replace", "replaceAll", "replaceChars", "replaceEach", "replaceEachRepeatedly",
+        "replaceFirst", "replaceIgnoreCase", "replaceOnce", "replaceOnceIgnoreCase",
+        "replacePattern", "reverse", "reverseDelimited", "right", "rightPad", "rotate", "split",
+        "splitByCharacterType", "splitByCharacterTypeCamelCase", "splitByWholeSeparator",
+        "splitByWholeSeparatorPreserveAllTokens", "splitPreserveAllTokens", "strip", "stripAccents",
+        "stripAll", "stripEnd", "stripStart", "stripToEmpty", "stripToNull", "substring",
+        "substringAfter", "substringAfterLast", "substringBefore", "substringBeforeLast",
+        "substringBetween", "substringsBetween", "swapCase", "toCodePoints", "toEncodedString",
+        "toRootLowerCase", "toRootUpperCase", "toString", "trim", "trimToEmpty", "trimToNull",
+        "truncate", "uncapitalize", "unwrap", "upperCase", "valueOf", "wrap", "wrapIfMissing"
+      ])
+  }
+
+  private predicate isExcludedParameter(int arg) {
+    this.getName().matches(["appendIfMissing%", "prependIfMissing%"]) and arg = [2, 3]
+    or
+    this.getName().matches(["remove%", "split%", "substring%", "strip%"]) and
+    arg = [1 .. getNumberOfParameters() - 1]
+    or
+    this.getName().matches(["chomp", "getBytes", "replace%", "toString", "unwrap"]) and arg = 1
+    or
+    this.getName() = "join" and
+    // Exclude joins of types that render numerically (char[] and non-primitive arrays
+    // are still considered taint sources)
+    this.getParameterType(arg).(Array).getComponentType() instanceof SafeJoinArrayType and
+    arg = 0
+  }
+
+  override predicate returnsTaintFrom(int arg) {
+    arg = [0 .. getNumberOfParameters() - 1] and
+    not exists(Type excluded | excluded = getAnExcludedParameterType() |
+      this.getParameterType(arg) = excluded
+    ) and
```

Comment:
Simplifies to:
```suggestion
    not this.getParameterType(arg) = getAnExcludedParameterType() and
```

Reply:
I don't think so -- is getParameterType(arg) is `int` for example then `not this.getParameterType(arg) = getAnExcludedParameterType()` can be satisfied by `getAnExcludedParameterType()` = `float`

Reply:
The variable that you've made explicit with an `exists` is still implicitly there in the simplified version, and its occurrence is still under the negation so the rewrite shouldn't change anything. Try it out for yourself. Here's e.g. a quick query for you:
```
int twoValues() { result = [1, 2] }

int getOne() { result = 1 }

predicate test1(string t) { t = "nope" and not exists(int n | getOne() = n | n = twoValues()) }

predicate test2(string t) { t = "nope" and not getOne() = twoValues() }
```
Both `test1` and `test2` have zero tuples.

Reply:
Also, recall that in QL all 3 of `=`, `in`, and `instanceof` essentially mean the same thing - they're all just syntactic sugar for the same underlying join operation.

Reply:
Doh, I wrote that exact test but put the multi-valued and single-valued arms of the conditional the wrong way around. Committed now.